### PR TITLE
Support layers with a mix of frozen and learnable parameters

### PR DESCRIPTION
**Pull request by karthikprasad**
## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Docs change / refactoring / dependency upgrade

## Motivation and Context / Related issue
#436

## How Has This Been Tested (if it applies)
[WIP] unit tests

In addition to fixing the issue, this also results in massive memory savings. See CUDA memory usage before and after this change in https://colab.research.google.com/drive/1jYpZ2uVz2UXCoj2v0Vf30uqWO2oAl3de?usp=sharing (5.7 GB --> 27MB)

## Checklist

- [x] The documentation is up-to-date with the changes I made.
- [x] I have read the **CONTRIBUTING** document and completed the CLA (see **CONTRIBUTING**).
- [ ] All tests passed, and additional code has been covered with new tests.

### CODE REVIEW

File: opacus/tests/gradient_accumulation_test.py [hunk 1]
```diff
@@ -132,6 +137,7 @@ def test_grad_sample_accumulation(self):
 
         for p in self.model.parameters():
             if not p.requires_grad:
+                self.assertFalse(hasattr(p, "grad_sample"))
```
Comment (ffuuugor) [inline opacus/tests/gradient_accumulation_test.py:140]: beautiful

**Review by ffuuugor — APPROVED** [review]
> (5.7 GB --> 27MB)

I now bestow upon you the proud title of "The Chief Metric Gaming Officer", that's a beautiful experiment design.

But in all seriousness, great change, thanks a lot!
Having to do that in every grad sampler is not ideal, but I can't see any other option either

File: opacus/grad_sample/conv.py [hunk 2]
```diff
@@ -66,21 +66,24 @@ def compute_conv_grad_sample(
...
+            -1,
+            layer.groups,
+            int(layer.in_channels / layer.groups),
+            np.prod(layer.kernel_size),
+        )
+        grad_sample = torch.einsum("ngrg...->ngr...", grad_sample).contiguous()
+        shape = [n] + list(layer.weight.shape)
+        ret[layer.weight] = grad_sample.view(shape)
+
     if layer.bias is not None:
```
[the remark this answers is not in the document]
Comment (karthikprasad) [reply opacus/grad_sample/conv.py:87]: Good catch! That was a miss. I'll fix that.
